File: server/src/main/java/dev/slimevr/posestreamer/PoseDataStream.java
package dev.slimevr.posestreamer;

import dev.slimevr.tracking.processor.skeleton.HumanSkeleton;

import java.io.*;


public abstract class PoseDataStream implements AutoCloseable {

	protected final OutputStream outputStream;
	protected boolean closed = false;

	protected PoseDataStream(OutputStream outputStream) {
		this.outputStream = outputStream;
	}

	protected PoseDataStream(File file) throws FileNotFoundException {
		this(new FileOutputStream(file));
	}

	protected PoseDataStream(String file) throws FileNotFoundException {
		this(new FileOutputStream(file));
	}

	public void writeHeader(HumanSkeleton skeleton, PoseStreamer streamer) throws IOException {
	}

	abstract void writeFrame(HumanSkeleton skeleton) throws IOException;

	public void writeFooter(HumanSkeleton skeleton) throws IOException {
	}

	public boolean isClosed() {
		return closed;
	}

	@Override
	public void close() throws IOException {
		outputStream.close();
		closed = true;
	}
}

File: server/core/src/main/java/dev/slimevr/posestreamer/PoseDataStream.kt
package dev.slimevr.posestreamer

import dev.slimevr.tracking.processor.skeleton.HumanSkeleton
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream
import java.lang.AutoCloseable

abstract class PoseDataStream protected constructor(protected val outputStream: OutputStream) : AutoCloseable {
	var isClosed: Boolean = false
		protected set

	protected constructor(file: File) : this(FileOutputStream(file))
	protected constructor(file: String) : this(FileOutputStream(file))

	@Throws(IOException::class)
	open fun writeHeader(skeleton: HumanSkeleton, streamer: PoseStreamer) {
	}

	@Throws(IOException::class)
	abstract fun writeFrame(skeleton: HumanSkeleton)

	@Throws(IOException::class)
	open fun writeFooter(skeleton: HumanSkeleton) {
	}

	@Throws(IOException::class)
	override fun close() {
		outputStream.close()
		this.isClosed = true
	}
}
